File: 2012/submissions/Khalil Fazal/1/Haskell/Main.hs
#! /usr/bin/env runghc

module Main where

import  Control.Monad               (ap, when)
import  Control.Monad.Instances     (Monad)
import  System.Environment          (getArgs)
import  System.IO                   (isEOF)

main :: IO ()
main = getArgs >>= ap (when . (> 0) . length) (continue . head)

continue :: String -> IO ()
continue input = do
    eof <- isEOF

    if eof then
        return ()
    else
        getLine >>= print . firstIndex input >> main

firstIndex :: Eq a => [a] -> [a] -> Int
firstIndex = firstIndex' 0

firstIndex' :: Eq a => Int -> [a] -> [a] -> Int
firstIndex' i word input | length input < 1 || i >= length input   = -1
                         | whileLoop i 0 word input == length word = i
                         | otherwise                               = firstIndex' (i + 1) word input

whileLoop :: Eq a => Int -> Int -> [a] -> [a] -> Int
whileLoop i j word input | j < length word && i + j < length input && input !! (i + j) == word !! j = whileLoop i (j + 1) word input
                         | otherwise                                                                = j

File: 2012/submissions/Khalil Fazal/1/Lua/Main.lua
#! /usr/bin/env lua

-- Lua sucks because arrays are indexed from 1
-- Lua sucks because Strings are "tables" and therefore its characters can't be easily accessed
-- Lua sucks because neither "j++" nor "j += 1" work, I can understand "j++" but "j += 1"?

function charAt(str, index)
    return string.sub(str, index + 1, index + 1)
end

function firstIndex(word, input)
    for i = 0, #input do
        j = 0

        while j < #word and i + j < #input and charAt(input, i + j) == charAt(word, j) do
            j = j + 1
        end

        if j == #word then
            return i
        end
    end

    return -1
end

if #arg > 0 then
    for input in io.lines() do
        print(firstIndex(arg[1], input))
    end
end
